### PR TITLE
allow adding plugins and change defaultmode for opensearch dashboards

### LINKED ISSUE #341: [Enhancement][opensearch-dashboards]Provide entry point for installing external plugins
**Is your feature request related to a problem? Please describe.**
Here is what I want to do: install an external plugin for Opensearch dashboard, but there is no entry/param to feed that. I figured to use post start lifecycle hooks to install plugins and was able to install plugins successfully that way, however to use the plugin, it also involves a schema change to add custom configs into the opensearch_dashboards.yml. I was not able to add additional configurations into the yaml file using the lifecycle hooks since the file was mounted as read only. 
I think there are 2 areas for improvement described below.

**Describe the solution you'd like**
1. Provide entry point to install external plugins
2. Provide entry point to allow changing permissions for config

**Describe alternatives you've considered**
The only other way to achieve that I think would work is to use a custom built image instead of using the official OSD image

**Additional context**
NA

### PR DESCRIPTION
**Pull request by BionIT**
Signed-off-by: Lu Yu <nluyu@amazon.com>

### Description
Similar to #71 , the change would allow installing external plugins when container starts for opensearch dashboards. 
In addition, this change allows to specify defaultMode for the opensearch_dashboards.yml file which is mounted as configmap
Changelog and Readme are updated.
Changes were tested using k8s cluster using the chart
 
### Issues Resolved
#341 
 
### Check List
- [x] Commits are signed per the DCO using --signoff

For any changes to files within Helm chart directories:
- [x] Helm chart version bumped
- [x] Helm chart `CHANGELOG.md` updated to reflect change

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by TheAlgo** [comment]
@BionIT Can you refactor the PR according to contribution guidelines mentioned in  : [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)

**Comment by TheAlgo** [comment]
@BionIT Can you refactor the PR according to contribution guidelines mentioned in  [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)

**Comment by TheAlgo** [comment]
@BionIT Can you refactor the PR according to contribution guidelines mentioned in  [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)

**Review by TheAlgo — DISMISSED** [review]
@BionIT Can you refactor the PR according to contribution guidelines mentioned in  [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)

**Comment by BionIT** [comment]
> @BionIT Can you refactor the PR according to contribution guidelines mentioned in [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)

@TheAlgo Sure! Do we need to bump `appVersion` from 2.3.0 to 2.3.1 as well or just `version`? The linked PR did both, but OSD latest release is at 2.3.0

**Comment by peterzhuamazon** [comment]
> > @BionIT Can you refactor the PR according to contribution guidelines mentioned in [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)
> 
> @TheAlgo Sure! Do we need to bump `appVersion` from 2.3.0 to 2.3.1 as well or just `version`? The linked PR did both, but OSD latest release is at 2.3.0

appversion keep 2.3.0, bump version.

**Comment by BionIT** [comment]
> > > @BionIT Can you refactor the PR according to contribution guidelines mentioned in [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)
> > 
> > 
> > @TheAlgo Sure! Do we need to bump `appVersion` from 2.3.0 to 2.3.1 as well or just `version`? The linked PR did both, but OSD latest release is at 2.3.0
> 
> appversion keep 2.3.0, just bump version.

Thanks @peterzhuamazon and @TheAlgo ! I have bumped version for osd chart, linter was complaining about version bump for os, but I only updated documentation for os chart, and per [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md): `Chart version should be bumped for every code change except docoumentation change`. Please let me know if any advice~

**Comment by TheAlgo** [comment]
> > > > @BionIT Can you refactor the PR according to contribution guidelines mentioned in [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)
> > > 
> > > 
> > > @TheAlgo Sure! Do we need to bump `appVersion` from 2.3.0 to 2.3.1 as well or just `version`? The linked PR did both, but OSD latest release is at 2.3.0
> > 
> > 
> > appversion keep 2.3.0, just bump version.
> 
> Thanks @peterzhuamazon and @TheAlgo ! I have bumped version for osd chart, linter was complaining about version bump for os, but I only updated documentation for os chart, and per [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md): `Chart version should be bumped for every code change except docoumentation change`. Please let me know if any advice~

Yes this is not ideal but this is a backlog for this repo. Our linter is configured this way. @prudhvigodithi / @peterzhuamazon We need to fix this ASAP. In the meanwhile @BionIT can you bump minor version for OS chart as well so that linter is happy

**Comment by BionIT** [comment]
> > > > > @BionIT Can you refactor the PR according to contribution guidelines mentioned in [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md)
> > > > 
> > > > 
> > > > @TheAlgo Sure! Do we need to bump `appVersion` from 2.3.0 to 2.3.1 as well or just `version`? The linked PR did both, but OSD latest release is at 2.3.0
> > > 
> > > 
> > > appversion keep 2.3.0, just bump version.
> > 
> > 
> > Thanks @peterzhuamazon and @TheAlgo ! I have bumped version for osd chart, linter was complaining about version bump for os, but I only updated documentation for os chart, and per [CONTRIBUTING.md](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md): `Chart version should be bumped for every code change except docoumentation change`. Please let me know if any advice~
> 
> Yes this is not ideal but this is a backlog for this repo. Our linter is configured this way. @prudhvigodithi / @peterzhuamazon We need to fix this ASAP. In the meanwhile @BionIT can you bump minor version for OS chart as well so that linter is happy

Sure, @TheAlgo! Thanks for clarifying! I have bumped version for os chart as well and the linter is happy now :).

File: charts/opensearch-dashboards/templates/deployment.yaml [hunk 1]
```diff
@@ -47,6 +47,9 @@ spec:
         - name: config
           configMap:
             name: {{ template "opensearch-dashboards.fullname" . }}-config
+            {{- if .Values.opensearchDashboardsYml.defaultMode }}
```
[the remark this answers is not in the document]
Comment (BionIT) [reply charts/opensearch-dashboards/templates/deployment.yaml:50]: There are multiple configmaps, and I think being explicit would help if we introduce more custom configurations about defaultmode for those configmaps. Open to a better naming if we have any

[the remark this answers is not in the document]
Comment (TheAlgo) [reply charts/opensearch-dashboards/templates/deployment.yaml:50]: @prudhvigodithi ^^

**Comment by peterzhuamazon** [comment]
Adding @prudhvigodithi to take a look at this.

**Comment by TheAlgo** [comment]
@prudhvigodithi @DandyDeveloper Please take a look when free. Thanks

**Comment by seanneumann** [comment]
Heya, what's the latest on getting this reviewed? I'd like to get this merged.

File: charts/opensearch-dashboards/templates/deployment.yaml [hunk 2]
```diff
@@ -151,6 +154,18 @@ spec:
...
+{{- if .Values.plugins.enabled }}
+        command:
+          - sh
+          - -c
+          - |
+            #!/usr/bin/bash
+            set -e
+            {{- range $plugin := .Values.plugins.installList }}
+            ./bin/opensearch-dashboards-plugin install {{ $plugin }}
+            {{- end }}
```
Comment (prudhvigodithi) [inline charts/opensearch-dashboards/templates/deployment.yaml:166]: Hey @BionIT quick question, for OpenSearch we have a `--batch` option to force install without a prompt, is there a same flag or scenario for `opensearch-dashboards-plugin`, I dont see it in the document though https://opensearch.org/docs/latest/dashboards/install/plugins/, please test this and share your update.

**Comment by prudhvigodithi** [comment]
Hey @BionIT thanks for the PR and contribution, please check the following points.

- Adding default mode is not a bad idea but can you share what issues faced for this to get added, if not lets keep what kubernetes sets to the defaultMode to the volume.
- Since the change is particular to the OpenSearch Dashboard, we dont have to increment the versions in chart.yaml for OpenSearch, only for the dashboard chart is required.
- Can you also please resolve the conflicts?

@TheAlgo @DandyDeveloper @peterzhuamazon @bbarani

File: charts/opensearch-dashboards/templates/deployment.yaml (hunk 2 above)
Comment (BionIT) [reply charts/opensearch-dashboards/templates/deployment.yaml:166]: Hey @prudhvigodithi, --batch is not supported by opensearch-dashboards-plugin(`error: unknown option '--batch'
`), I have been using this script to install plugins and never had prompts.

**Comment by BionIT** [comment]
> Hey @BionIT thanks for the PR and contribution, please check the following points.
> 
> * Adding default mode is not a bad idea but can you share what issues faced for this to get added, if not lets keep what kubernetes sets to the defaultMode to the volume.
>   ~- Since the change is particular to the OpenSearch Dashboard, we dont have to increment the versions in chart.yaml for OpenSearch, only for the dashboard chart is required.~
> * Can you also please resolve the conflicts?
> 
> @TheAlgo @DandyDeveloper @peterzhuamazon @bbarani @seraphjiang

Per request, resolved conflicts.
Per allowing users to set defaultMode, k8s allow users to specify defaultmode for volumes, why our helm chart want to restrict it to readonly? We have use case to dynamically change the values in the yaml and the current set up prevents that.

**Comment by peterzhuamazon** [comment]
Pinging @DandyDeveloper @TheAlgo @prudhvigodithi again for the checks on @BionIT latest changes.
I am ok with this approach, and if any of you give an approval we can follow up with a merge.

Thanks.

**Review by rishabh6788 — APPROVED** [review]

**Review by peterzhuamazon — APPROVED** [review]
Approved, thanks for being patient on this.
@rishabh6788 please approve.

Thanks.

**Comment by peterzhuamazon** [comment]
Related to kind action need to update to 1.3.0.

**Comment by peterzhuamazon** [comment]
Hi @BionIT please have a follow-up backport PR to 1.x.
Thanks.

**Comment by peterzhuamazon** [comment]
Backport 1.x: #359

**Pull request merged**